api/rosetta/errors.go

```diff
@@ -33,94 +32,106 @@ type Error struct {
-			"blockchain": fail.Blockchain,
-			"network":    fail.Network,
-		},
-	}
+	return newError(configuration.ErrorInvalidNetwork, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func InvalidAccount(fail failure.InvalidAccount) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidAccount,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"address": fail.Address,
-			"chain":   fail.Chain,
-		},
-	}
+	return newError(configuration.ErrorInvalidAccount, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func InvalidCurrency(fail failure.InvalidCurrency) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidCurrency,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"symbol":   fail.Symbol,
-			"decimals": fail.Decimals,
-		},
-	}
+	return newError(configuration.ErrorInvalidCurrency, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func InvalidBlock(fail failure.InvalidBlock) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidBlock,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"index": fail.Index,
-			"hash":  fail.Hash,
-		},
-	}
+	return newError(configuration.ErrorInvalidBlock, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func InvalidTransaction(fail failure.InvalidTransaction) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorInvalidTransaction,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"hash": fail.Hash,
-		},
-	}
+	return newError(configuration.ErrorInvalidTransaction, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func UnknownCurrency(fail failure.UnknownCurrency) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorUnknownCurrency,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"symbol":   fail.Symbol,
-			"decimals": fail.Decimals,
-		},
-	}
+	return newError(configuration.ErrorUnknownCurrency, fail.Message, optsFromMap(fail.Params)...)
 }
 
 func UnknownBlock(fail failure.UnknownBlock) Error {
-	return Error{
-		ErrorDefinition: configuration.ErrorUnknownBlock,
-		Description:     fail.Message,
-		Details: map[string]interface{}{
-			"index": fail.Index,
-			"hash":  fail.Hash,
-		},
+	return newError(configuration.ErrorUnknownBlock, fail.Message, optsFromMap(fail.Params)...)
+}
+
+func optsFromMap(m map[string]interface{}) []OptionFunc {
+
+	opts := make([]OptionFunc, 0, len(m))
+	for key, value := range m {
+		opts = append(opts, With(key, value))
+	}
+
+	return opts
+}
+
+func WithInt(key string, value int) OptionFunc {
+	return func(err *Error) {
+		err.Details[key] = value
+	}
+}
+
+func WithString(key string, value string) OptionFunc {
+	return func(err *Error) {
+		err.Details[key] = value
+	}
+}
+
+// TODO: check - silently occupy 'error' field ok?
+func WithError(e error) OptionFunc {
+	return func(err *Error) {
+		err.Details["error"] = e.Error()
+	}
+}
+
+func With(key string, value interface{}) OptionFunc {
```

Comment:
I ended up adding the generic version too, since we're adding fields here that were populated in other packages (e.g. validator). Fields we're currently adding are integer variants (`int`, `uint`, `uint64`) and strings, so we should not have any marshalling issues.

Flow identifiers and similar are already string encoded when setting the parameters, but we could always handle them explicitely here, just in case.